feat: support Nitro enclave

GnatorX opened the pull request:
Fixes #3693
and https://github.com/aws/karpenter-provider-aws/issues/7047


**Description**

This doesn't solve the issue of resources carved out for the enclave, it simply allows Karpenter to launch instances with enclave enabled via Launch Template.

https://docs.aws.amazon.com/enclaves/latest/user/kubernetes.html#create-lt

Follow this doc, you specify how much CPU and memory is carved out for the enclave making it more in the realm of Node Overlay for resource modification.

It also requires huge pages support for the memory carved out which also isn't supported by Karpenter atm. 

We are only able to use this with Custom AMI because system reserves aren't copied to the underlying node. What we do here is carve out the memory and CPU via system reserve in ec2nodeclass.

**How was this change tested?**
- Created a nodeclass that launched a node with enclave enabled

Karpenter logs:
```
NodeClaim: {
     name: 
   }
   allocatable: {
     cpu: 3920m
     ephemeral-storage: 89Gi
     memory: 14162Mi
     pods: 110
     vpc.amazonaws.com/pod-eni: 18
   }
   capacity-type: on-demand
   controller: nodeclaim.lifecycle
   controllerGroup: karpenter.sh
   controllerKind: NodeClaim
   instance-type: m5.xlarge
   level: INFO
   logger: controller
   message: launched nodeclaim
   name: 
   namespace:
   provider-id: aws:///us-west-2a/i-03da14a00d67b7a62
   reconcileID: 
   time: 
   zone: 
}
```

![Screenshot 2025-07-01 at 3 50 20 PM](https://github.com/user-attachments/assets/0063fc55-7336-4f70-9705-12d59dbb38fd)


**Does this change impact docs?**
- [x] Yes, PR includes docs updates <!-- docs must be added to /preview to be included in future version releases -->
- [ ] Yes, issue opened: # <!-- issue number -->
- [] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

GnatorX committed:
Update karpenter.k8s.aws_ec2nodeclasses.yaml

GnatorX committed:
Update ec2nodeclass.go

GnatorX committed:
Update zz_generated.deepcopy.go

GnatorX committed:
Update resolver.go

GnatorX committed:
Update launchtemplate.go

GnatorX committed:
Update types.go

GnatorX commented:
Updated with more context around what enclave requires

GnatorX committed:
Merge branch 'main' into nitro-enclave

GnatorX committed:
Update karpenter.k8s.aws_ec2nodeclasses.yaml

GnatorX committed:
Update resolver.go

GnatorX committed:
Merge branch 'main' into nitro-enclave

GnatorX committed:
Merge branch 'main' into nitro-enclave

GnatorX committed:
Merge branch 'main' into nitro-enclave

garvinp-stripe committed:
add unit test

garvinp-stripe committed:
add hash test

GnatorX committed:
Merge branch 'main' into nitro-enclave

garvinp-stripe committed:
doc update

garvinp-stripe committed:
add info on how to remove resources from karpenter simulation

GnatorX commented on pkg/apis/crds/karpenter.k8s.aws_ec2nodeclasses.yaml:
```diff
@@ -293,6 +293,18 @@ spec:
                 detailedMonitoring:
                   description: DetailedMonitoring controls if detailed monitoring is enabled for instances that are launched
                   type: boolean
+                enclaveOptions:
```
doc updated

GnatorX committed:
Merge branch 'main' into nitro-enclave

GnatorX commented on pkg/apis/v1/ec2nodeclass.go:
```diff
@@ -138,6 +138,11 @@ type EC2NodeClassSpec struct {
+	// If omitted, defaults to false.
+	// +kubebuilder:default={"enabled":false}
+	// +optional
+	EnclaveOptions *EnclaveOptions `json:"enclaveOptions,omitempty"`
```
added a hash test

GnatorX committed:
fix unit test, remove default to avoid kicking off drift

GnatorX committed:
fix test

GnatorX committed:
remove default

GnatorX committed:
Merge branch 'main' into nitro-enclave

GnatorX committed:
Merge branch 'main' into nitro-enclave

GnatorX committed:
Merge branch 'aws:main' into nitro-enclave

GnatorX commented:
I believe we are waiting on node overlay https://github.com/kubernetes-sigs/karpenter/pull/2166 @engedaam can we unblock this or are we waiting on node overlay implementation?

GnatorX committed:
Merge branch 'aws:main' into nitro-enclave

GnatorX committed:
Merge branch 'aws:main' into nitro-enclave

GnatorX committed:
Merge branch 'aws:main' into nitro-enclave

engedaam commented on website/content/en/docs/concepts/nodeclasses.md:
```diff
@@ -1528,6 +1532,142 @@ If a `NodeClaim` requests `vpc.amazonaws.com/efa` resources, `spec.associatePubl
+
+To use Nitro Enclaves, you must use the Custom AMI family to remove system resources for the enclave. Currently, this can only be achieved with the Custom AMI family combined with systemReserve to remove resources, since the Custom AMI family doesn't translate the system reserve to the underlying node but can remove the resources from Karpenter's scheduling simulation.
+
+### What are Nitro Enclaves?
```
Let's add a separate task section for the enclaves, so that we can describe the configuration that is need for setting up HugePages with NodeOverlay and the userData.

engedaam commented on pkg/providers/amifamily/resolver.go:
```diff
@@ -306,6 +307,7 @@ func (r DefaultResolver) resolveLaunchTemplates(
 			CapacityType:            capacityType,
 			CapacityReservationID:   id,
 			CapacityReservationType: capacityReservationType,
+			EnclaveOptions:          nodeClass.Spec.EnclaveOptions,
```
Since nitro enclaves are not support for all instance types, do we need to make separate launch templates for instance types that support nitro enclaves vs once that do not support the feature? https://docs.aws.amazon.com/enclaves/latest/user/nitro-enclave.html

DerekFrank replied:
It does seem like an easy switch later. If we start without separate launch templates and just solid documentation, we can switch to separate launch templates later as folks need that support

engedaam commented on pkg/providers/launchtemplate/suite_test.go:
```diff
@@ -2602,6 +2603,132 @@ eviction-max-pod-grace-period = 10
 		Entry("enabled", true),
 		Entry("disabled", false),
 	)
+	Context("Enclave Options", func() {
```
Let's add an integration test to make sure that the instances are launched as expected with enclaves

garvinp-stripe replied:
how do i add an integration test?

GnatorX commented on pkg/apis/v1/ec2nodeclass.go:
```diff
@@ -139,6 +139,10 @@ type EC2NodeClassSpec struct {
 	// +kubebuilder:default={"httpEndpoint":"enabled","httpProtocolIPv6":"disabled","httpPutResponseHopLimit":1,"httpTokens":"required"}
 	// +optional
 	MetadataOptions *MetadataOptions `json:"metadataOptions,omitempty"`
+	// EnclaveOptions specifies whether the instance is enabled for Amazon Web Services Nitro Enclaves.
```
I think users usually rely on the device plugin so I didn't think we need a label(?)

xiaoen-liu commented:
Is there any further progress expected on this PR?
It looks like this PR, combined with the NodeOverlay feature, would enable full enclave support.
@engedaam @GnatorX

garvinp-stripe committed:
update nitro docs

garvinp-stripe committed:
Merge branch 'main' into nitro-enclave

GnatorX committed:
Merge branch 'main' into nitro-enclave

DerekFrank commented on website/content/en/docs/tasks/nitro-enclaves.md:
````diff
@@ -0,0 +1,205 @@
+    consolidateAfter: 1m
+```
+
+## Additional Resources
````
Are we missing a validation step here? How do I tell it worked?

GnatorX replied:
added quick validation

GnatorX commented on website/content/en/docs/tasks/nitro-enclaves.md:
```diff
@@ -0,0 +1,205 @@
+  enclaveOptions:
+    enabled: true
+  
+  # Select your AMI family - AL2023 is recommended
```
I chose Al2023 because that is the most common in Karpenter. I can remove the recommendation

GnatorX commented on website/content/en/docs/tasks/nitro-enclaves.md:
```diff
@@ -0,0 +1,205 @@
+  name: nitro-enclaves-overlay
+spec:
+  # Select instance types that support Nitro Enclaves
+  requirements:
```
ya

garvinp-stripe committed:
update with int test and doc update

GnatorX committed:
Merge pull request #2 from GnatorX/garvin/nitro-enclave

Update with lint and integration test

GnatorX committed:
Update note on Karpenter and Nitro Enclaves

Clarify behavior of Karpenter when enabling enclave options and mention separate nodeclasses for Nitro Enclave instances.

GnatorX commented on website/content/en/docs/tasks/nitro-enclaves.md:
```diff
@@ -0,0 +1,205 @@
+        - key: karpenter.k8s.aws/instance-family
+          operator: In
+          values:
+            - m5
```
updated it to include instance size skipping `large`

GnatorX committed:
Merge branch 'aws:main' into nitro-enclave

DerekFrank reviewed (commented):
/karpenter snapshot

GnatorX committed:
Merge branch 'main' into nitro-enclave

GnatorX committed:
Merge branch 'main' into nitro-enclave

abbas-square commented:
@DerekFrank is there any update?

GnatorX committed:
Merge branch 'aws:main' into nitro-enclave

GnatorX commented:
We chatted a bit offline (between AWS and myself), and thank you folks from AWS (Derek and Alex) for engaging on this and helping me with throwing ideas and getting changes into Karpenter.  It seems like AWS is interested in changing how nitro enclaves are configured in the underlying AWS API and it seems like they wish to hold this PR till the new API is defined to reduce sprawl of Karpenter's API.

Although I do understand this desire, given the communities' desire for this feature I am wondering if it make sense to merge this PR for the short/ medium term. Given that I don't think the existing nitro enclave API would be deprecated, no timeline for the new API and a lot of users that wants this feature, I think its reasonable to compromise the succinctness of Karpenter's API to meet users need.

yan commented:
Thanks for following up, Garvin ! Would definitely hugely appreciate support for Nitro as it is today, if at all possible.

GnatorX committed:
Merge branch 'main' into nitro-enclave

amlozano1 commented:
@akestner Do you have any more details at this time?
